dash/include/dash/halo/Halo.h

```diff
@@ -0,0 +1,1104 @@
+    }
+
+    return index;
+  }
+
+ private:
+  void pushBndElems( dim_t dim, std::array<index_type, NumDimensions>& offsets,
+                     std::array<size_type, NumDimensions>& extents,
+                     const HaloExtsMaxT&                   halo_exts_max,
+                     const CycleSpecT&                     cycle_spec ) {
+    for( auto d_tmp = dim + 1; d_tmp < NumDimensions; ++d_tmp ) {
+      if( cycle_spec[d_tmp] == Cycle::NONE ) {
+        if( offsets[d_tmp] < halo_exts_max[d_tmp].first ) {
+          offsets[d_tmp] = halo_exts_max[d_tmp].first;
+          extents[d_tmp] -= halo_exts_max[d_tmp].first;
+          if( ( offsets[d_tmp] + extents[d_tmp] + halo_exts_max[d_tmp].second )
+              > _pattern.extent( d_tmp ) )
+            extents[d_tmp] -= halo_exts_max[d_tmp].second;
+          continue;
+        }
+        auto check_extent_tmp =
+          offsets[d_tmp] + extents[d_tmp] + halo_exts_max[d_tmp].second;
+        if( check_extent_tmp > _pattern.extent( d_tmp ) ) {
+          extents[d_tmp] -= halo_exts_max[d_tmp].second;
+        }
+      }
+    }
+    ViewSpecT boundary_next( offsets, extents );
+    _size_bnd_elems += boundary_next.size();
+    _boundary_elements.push_back( std::move( boundary_next ) );
+  }
+
+ private:
+  GlobMemT& _globmem;
+
+  const PatternT& _pattern;
+
+  const ViewSpecT& _view;
+
+  const halo_reg_spec_t& _halo_reg_spec;
+
+  ViewSpecT _view_safe;
+
+  ViewSpecT _view_inner;
+
+  regions_t _halo_regions;
+
+  std::array<region_t*, HaloRegionSpecT::MaxIndex> _halo_reg_mapping{};
+
+  regions_t _boundary_regions;
+
+  std::array<region_t*, HaloRegionSpecT::MaxIndex> _boundary_reg_mapping{};
+
+  std::vector<ViewSpecT> _boundary_elements;
+
+  size_type _size_bnd_elems = 0;
+
+};  // class HaloBlock
+
+template <typename HaloBlockT>
+class HaloMemory {
+ private:
+  static constexpr dim_t NumDimensions = HaloBlockT::ndim();
+  using HaloRegionSpecT                = HaloRegionSpec<NumDimensions>;
+  using region_index_t = typename HaloRegionSpecT::region_index_t;
+  static constexpr region_index_t MaxIndex = HaloRegionSpecT::MaxIndex;
+  using PatternT                           = typename HaloBlockT::pattern_t;
+  using index_type                         = typename PatternT::index_type;
+  using size_type                          = typename PatternT::size_type;
+
+  static constexpr MemArrange MemoryArrange = PatternT::memory_order();
+
+ public:
+  using ElementT = typename HaloBlockT::value_t;
+  using CoordsT  = std::array<index_type, NumDimensions>;
+
+  HaloMemory( const HaloBlockT& haloblock ) : _haloblock( haloblock ) {
+    _halobuffer.resize( haloblock.halo_size() );
+    auto* offset = _halobuffer.data();
+
+    for( const auto& region : haloblock.halo_regions() ) {
+      _halo_offsets[region.index()] = offset;
+      offset += region.size();
+    }
+  }
+
+  ElementT* haloPos( region_index_t index ) { return _halo_offsets[index]; }
```

Comment:
Also: you're using both underscores and CamelCase in the public interface, I think we should consolidate that to underscores.